### PR TITLE
SLS-2493: Wrap top level spans with gcp.cloudrun

### PR DESCRIPTION
**Pull request by IvanTopolcic**
### What does this PR do?

This PR adds a `modifyTrace` middleware function where we process traces and send them to trace intake. This is requried for serverless as we need to check each cloudrun request for top level spans and wrap them in a gcp.cloudrun span.

image

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/serverless/trace/trace_modifier.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		Start:    oldRoot.Start,
+		SpanID:   rand.Uint64(),
+		Duration: oldRoot.Duration,
+		Error:    oldRoot.Error,
+		Meta:     oldRoot.Meta,
+		Type:     oldRoot.Type,
+		Service:  oldRoot.Service,
+	}
+	oldRoot.ParentID = root.SpanID
+	*trace = append(*trace, root)
```
Comment (maxday) [inline pkg/serverless/trace/trace_modifier.go:36]: I am a bit concerned about editing this in place if this data structure is not tread safe?,
Would it be possible to return a new array instead so we'll stick with a pure function?

File: pkg/trace/agent/agent.go [hunk 2]
```diff
@@ -60,6 +60,9 @@ type Agent struct {
 	// ModifySpan will be called on all spans, if non-nil.
 	ModifySpan func(*pb.Span)
 
+	// ModifyTraceFunc will be called on all traces, if non-nil.
+	ModifyTraceFunc func(*[]*pb.Span)
```
Comment (maxday) [inline pkg/trace/agent/agent.go:64]: nit: to be consistent with `ModifySpan` let's name it `ModifyTrace`? or rename `ModifySpan` to `ModifySpanFunc`?

Comment (IvanTopolcic) [reply pkg/trace/agent/agent.go]: Done!

**Review by maxday — APPROVED** [review]

File: pkg/serverless/trace/trace_modifier.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		Resource: "gcp.cloudrun",
+		Start:    oldRoot.Start,
+		SpanID:   rand.Uint64(),
+		Duration: oldRoot.Duration,
+		Error:    oldRoot.Error,
+		Meta:     oldRoot.Meta,
+		Type:     oldRoot.Type,
+		Service:  oldRoot.Service,
+	}
+	oldRoot.ParentID = root.SpanID
```
Comment (knusbaum) [inline pkg/serverless/trace/trace_modifier.go:35]: Are we sure this is ok? I'm currently asking backend if there's any case where SpanID == TraceID is important.

**Review by knusbaum — CHANGES_REQUESTED** [review]
See inline comments

**Pull request closed without merging**